tools/cndump/cndb_reader.c

```diff
@@ -0,0 +1,63 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/*
+ * Copyright (C) 2022 Micron Technology, Inc.  All rights reserved.
+ */
+
+#include <mpool/mpool.h>
+
+#include <hse_ikvdb/diag_kvdb.h>
+
+#include "cndb_reader.h"
+#include "cndb_record.h"
+#include "fatal.h"
+
+static void
+read_record(struct cndb_dump_reader *reader, struct cndb_rec *rec)
+{
+    merr_t err;
+    size_t reclen;
+
+    err = mpool_mdc_read(reader->mdc, rec->buf, rec->bufsz, &reclen);
+    if (merr_errno(err) == EOVERFLOW) {
+        cndb_rec_resize(rec, reclen);
+        err = mpool_mdc_read(reader->mdc, rec->buf, rec->bufsz, &reclen);
+    }
+
+    if (err)
+        fatal("mpool_mdc_read", err);
```

Comment:
The original error could be from deep in the HSE library.  This give me a hint about where in my code the call was made.  I could alternatively change fatal to be a macro and use `__FILE__, __LINE__`.

Reply:
That would probably be better. This could be an opportunity to use the static logging library which would handle this for you.

Reply:
I don't understand the resistance to naming the API that failed - especially given that this is a development tool.  This reporting style is common with with `sterror` and `perror`, for example:

https://en.cppreference.com/w/c/io/perror

Then I can do `man fopen` to see what the specific error code means for that API.

Reply:
If I call a function and it fails I prefer to name it in the error message, otherwise it's just a vague error message with insufficient context to understand what it means, which then requires the error message print out the file and line and a programmer to have to consult the code.  Customer is left scratching head...
The worse I've seen in HSE are all the "cannot open" error messages in ikvdb_open(), OMG, WTF???  Completely useless without the code...

Reply:
I don't really see what hardcoding function names gets you over using `__func__`, `__LINENO__`, and `__FILE__` at the point of the failure.

Reply:
If you tell me you got an out of memory error at line 456 in func foo() I probably won't know what it means and would definitely have to consult the code.  And these line numbers change all the time...

Buf if you tell me you got an out of memory error trying to create a route map entry in func foo I might be able to help without consulting the code.  And if everyone is hitting that problem it's easy to correlate and remember.
Line and func are a necessity for finding the call site, but having more context about the error is something that devs and support people can remember and work with, and might be able to suggest a know solution to "that" problem.

Reply:
Spoken like someone who's had to support enterprise storage SW in the field  :-)

Reply:
Once or twice ;-)  When I was new at Veritias I was in a mtg where they were trying to figure out how to fairly distribute pager duty.  Having not written a single line of code and NOT wanting pager duty, I suggested it should be proportional to the amount of code one has written.  Needless to say, there was an immediate NO from several of the people in the group, but it was all in good fun.  I ended up having to carry that thing, made no sense to me, but fortunately nothing ever happened on my watch.

Reply:
> Buf if you tell me you got an out of memory error trying to create a route map entry in func foo I might be able to help without consulting the code.

This seems to be an argument for logging in mpool_mdc_read() and not logging the function name here.

```c
fatal("mpool_mdc_read", err);
// mpool_mdc_read: <merr_t as string>

vs

log_errx("Failed to read the mpool MDC of the KVDB (%s)", err, kvdb_home);
// __FILE__:__LINE__: __func__: Failed to read the mpool MDC of the KVDB (%s): <merr_t as string>
```

I am not seeing the value of the first over the second.

Reply:
Core/library functions should never never never log errors, not only is there insufficient context for an informative message it could be called from a thousand places, in varied contexts (interrupt handlers, signal handlers, locks held, ...), just do what you were asked to do and nothing more. The right place to log an error is where you have enough context to log a meaningful message it's safe to call your logging function.
IMO this fatal("some-func", err); could be much more useful if it provided additional context, as-is I don't think it's very helpful to anyone not actively working on the code base...

Reply:
>  I don't think it's very helpful to anyone not actively working on the code base...

Agree.  But that's who will be using this tool.